### DISCUSSION TITLE
Split React components?

### DISCUSSION THREAD

**Discussion by astrojuanlu** [kedro-org/kedro-viz#1943] [Ideas]
In discussion yesterday with @jitu5 and @stephkaiser, the idea of breaking down the monolithic Kedro Viz React component came up. @jitu5 mentioned that it would be a bit of work.

How feasible it would be?

Some resources I found after a quick search:

- "Common React Mistakes: Monolithic Components and a Lack of Abstraction" https://www.pmg.com/blog/common-react-mistakes-monolithic-components-lack-abstraction
- Advantages of smaller components https://stackoverflow.com/q/49591923
- Disadvantages of big components https://stackoverflow.com/q/39391195
- React components composition
  - https://zhenyong.github.io/react/docs/multiple-components.html
  - https://www.developerway.com/posts/components-composition-how-to-get-it-right
- Sharing State Between Components https://react.dev/learn/sharing-state-between-components
  - ...using Redux https://stackoverflow.com/q/57323921
- How to Organize a Large React Application and Make It Scale https://www.sitepoint.com/organize-large-react-application/

**Comment by yetudada** [comment]
To guide this, we’re also meant to think of what it should achieve or potentially, what it should enable for our users so that it’s not an endless refactor. This discussion has come up before when we looked at making the flowchart a standalone component: https://github.com/kedro-org/kedro-viz/issues/954 & https://github.com/kedro-org/kedro-viz/issues/943.

**Reply by astrojuanlu** [reply]
Thanks @yetudada for the context. I appreciate that there's a log of the discussions, but very limited: @idanov asked in https://github.com/kedro-org/kedro-viz/issues/943#issuecomment-1385189879 whether we could do this split, then it was decided in a meeting that it wouldn't be done https://github.com/kedro-org/kedro-viz/issues/943#issuecomment-1385402852, but it's not really spelled out why.

**Comment by jitu5** [comment]
Extracting the flowchart component requires a considerable amount of refactoring to decouple it from the existing Kedro-Viz. This involves removing or replacing Redux with a lighter state management, which is a big task given the complexity and maturity of the current setup.

Above discussion came from feature request by @datajoely to have preview mode which is flowchart only mode. which can be achieve with react props. https://github.com/kedro-org/kedro-viz/issues/1745#issuecomment-2166402190 without going into major refactor.

But Standalone flowchart component going to be really helpful in use cases VSCode extension if we can run web/react app inside VSCode.

**Reply by astrojuanlu** [reply]
> But Standalone flowchart component going to be really helpful in use cases VSCode extension if we can run web/react app inside VSCode.

Indeed. From my understanding, if we ever want to "debundle" the UI components of Kedro-Viz, currently our only option is to embed identical React components, and selectively disable the elements we don't want.

**Comment by astrojuanlu** [comment]
Alternatives to Redux:

- React Contexts https://react.dev/learn/passing-data-deeply-with-context

> Usually, you will pass information from a parent component to a child component via props. But passing props can become verbose and inconvenient if you have to pass them through many components in the middle, or if many components in your app need the same information. Context lets the parent component make some information available to any component in the tree below it—no matter how deep—without passing it explicitly through props.

- Zustand https://www.npmjs.com/package/zustand
- Jotai https://www.npmjs.com/package/jotai (what [Meta started using instead of Recoil](https://github.com/facebook/sapling/commit/547b205eab16fc78d73ec8edb38b2b2bdc84ddf2))

**Reply by RojanSr** [reply]
Zustand is my personal favorite. Easy to use and understand.

**Comment by astrojuanlu** [comment]
TIL: We do have those individual components (thanks @rashidakanchwala)

https://github.com/kedro-org/kedro-viz/blob/4005b58e97da15244f0679bc50abbd09c95f528c/src/components/flowchart-wrapper/flowchart-wrapper.js#L322-L354

Is it technically possible for a user to do just

```javascript
<FlowChart />
```

? (Even if it's not currently documented and props come from Redux)

**Reply by datajoely** [reply]
this would be super powerful
